get_fsm: treat a timeout value of 0 as infinity

hmmr opened the pull request:
RTS-1328 (#1464)

When a client calls a `get` with a value of 0 for its `Timeout` parameter, get_fsm treats that value literally and immediately triggers a timeout (https://github.com/basho/riak_kv/blob/riak_ts-develop/src/riak_kv_get_fsm.erl#L544). One can argue that this is indeed expected behaviour, but the FSM still manages to serve some `get`s (which becomes possible under certain racy conditions between state changes and message passing to self), and that can be really confusing for the client.

This PR explicitly treats the case of `Timeout = 0` the same as `Timeout = infinity`, for the `get` API call.

andytill commented:
Suggest validating that the input is a positive integer, as per the current dialyzer spec. This enforces the current behaviour rather than changing it.

hmmr commented:
@andytill A zero timeout is allowed for other calls (delete, list_keys). Disallowing 0 for the get timeout would make it an odd one out.

Due to apparently different ways zero timeout is treated in those other calls, more extensive work is needed to make it uniformly mean infinity, which this PR does not claim to do.

andytill commented:
This change is making it the odd one out though, by rewriting the timeout.

It is just strange to rewrite the timeout because zero is the exact opposite of infinity.

Commit:
treat a timeout value of 0 as infinity in get, delete, list_keys

Failure to do so will confuse get_fsm when riak_client:get is called with
a Timeout of 0.

hmmr commented:
Here's another take. Zeros are now converted to `'infinity'` in riak_kv_ts_svc (i.e., underlying functions don't see 0 at all), for all relevant API calls (get, delete, list_keys).

jvoegele commented:
Just a gut reaction here, but treating `0` as `infinity` feels a bit too much like magic to me. I’d prefer more explicit behavior, such as disallowing `0`.

hmmr commented:
@jvoegele That's fair, but how do you specify `infinity` then?

jvoegele commented:
With the atom `infinity`? Maybe I'm missing something here...

andytill commented:
Currently `infinity` doesn't get passed to the fsm by riak_client, it only passes the given or default timeout.

hmmr commented:
The protobuff messages for the affected calls have an uint32 for it.

On 23 Aug 2016 17:27, "Jason Voegele" notifications@github.com wrote:

> With the atom infinity? Maybe I'm missing something here...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/basho/riak_kv/pull/1466#issuecomment-241748862, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/ABDkh3ng9C1r_oCEtC8Er1o1eZgK0NQpks5qiwNDgaJpZM4JqYBB
> .

hmmr commented:
A decision was reached to disallow 0 _and_ `;infinity'` as valid timeout values.
